x/posts/keeper/invariants.go

```diff
@@ -0,0 +1,222 @@
+				invalid = true
+			}
+
+			// Validate attachment
+			err := attachment.Validate()
+			if err != nil {
+				invalid = true
+			}
+
+			if invalid {
+				invalidAttachments = append(invalidAttachments, attachment)
+			}
+
+			return false
+		})
+
+		return sdk.FormatInvariant(types.ModuleName, "invalid attachments",
+			fmt.Sprintf("the following attachments are invalid:\n %s", formatOutputAttachments(invalidAttachments)),
+		), invalidAttachments != nil
+	}
+}
+
+// formatOutputAttachments concatenates the given attachment information into a string
+func formatOutputAttachments(attachments []types.Attachment) (output string) {
+	for _, attachment := range attachments {
+		output += fmt.Sprintf("SubspaceID: %d, PostID: %d, AttachmentID: %d\n",
+			attachment.SubspaceID, attachment.PostID, attachment.ID)
+	}
+	return output
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidUserAnswersInvariant checks that all the user answers are valid
+func ValidUserAnswersInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (message string, broken bool) {
+		var invalidUserAnswers []types.UserAnswer
+		k.IterateUserAnswers(ctx, func(_ int64, answer types.UserAnswer) (stop bool) {
+			invalid := false
+
+			// Check subspace
+			if !k.HasSubspace(ctx, answer.SubspaceID) {
+				invalid = true
+			}
+
+			// Check post
+			if !k.HasPost(ctx, answer.SubspaceID, answer.PostID) {
+				invalid = true
+			}
+
+			// Check the associated poll
+			if !k.HasPoll(ctx, answer.SubspaceID, answer.PostID, answer.PollID) {
+				invalid = true
+			}
+
+			// Validate the answer
+			err := answer.Validate()
+			if err != nil {
+				invalid = true
+			}
+
+			if invalid {
+				invalidUserAnswers = append(invalidUserAnswers, answer)
+			}
+
+			return false
+		})
+
+		return sdk.FormatInvariant(types.ModuleName, "invalid user answers",
+			fmt.Sprintf("the following user answers are invalid:\n %s", formatOutputUserAnswers(invalidUserAnswers)),
+		), invalidUserAnswers != nil
+	}
+}
+
+// formatOutputUserAnswers concatenates the given user answers information into a string
+func formatOutputUserAnswers(answers []types.UserAnswer) (output string) {
+	for _, answer := range answers {
+		output += fmt.Sprintf("SubspaceID: %d, PostID: %d, PollID: %d, User: %s\n",
+			answer.SubspaceID, answer.PostID, answer.PollID, answer.User)
+	}
+	return output
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidActivePollsInvariant checks that all the active polls are valid
+func ValidActivePollsInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (message string, broken bool) {
+		var invalidActivePolls []types.Attachment
+		k.IterateActivePolls(ctx, func(_ int64, attachment types.Attachment) (stop bool) {
+			poll := attachment.Content.GetCachedValue().(*types.Poll)
+
```

Comment:
This is not needed. The validation of the poll is already performed inside the `ValidAttachmentsInvariant`. This will iterate over polls anyway so no need to double check for validity here since the active poll queue only stores a reference to the polls and not different poll objects from the ones that are iterated over within the `ValidAttachmentsInvariant`